test/tls-provider.c

```diff
@@ -3258,23 +3258,29 @@ int tls_provider_init(const OSSL_CORE_HANDLE *handle,
      */
     if (!c_obj_create(handle, XORSIGALG_OID, XORSIGALG_NAME, XORSIGALG_NAME)) {
         ERR_raise(ERR_LIB_USER, XORPROV_R_OBJ_CREATE_ERR);
-        return 0;
+        goto err;
     }
 
     if (!c_obj_add_sigid(handle, XORSIGALG_OID, "", XORSIGALG_OID)) {
         ERR_raise(ERR_LIB_USER, XORPROV_R_OBJ_CREATE_ERR);
-        return 0;
+        goto err;
     }
     if (!c_obj_create(handle, XORSIGALG_HASH_OID, XORSIGALG_HASH_NAME, NULL)) {
         ERR_raise(ERR_LIB_USER, XORPROV_R_OBJ_CREATE_ERR);
-        return 0;
+        goto err;
     }
 
     if (!c_obj_add_sigid(handle, XORSIGALG_HASH_OID, XORSIGALG_HASH, XORSIGALG_HASH_OID)) {
         ERR_raise(ERR_LIB_USER, XORPROV_R_OBJ_CREATE_ERR);
-        return 0;
+        goto err;
     }
 
     *out = tls_prov_dispatch_table;
     return 1;
+
+err:
+    OPENSSL_free(prov_ctx);
+    *provctx = NULL;
```

Comment:
Err, should this be `provctx = NULL` (without the `*`)? If `OPENSSL_free()` works the same as `free()`, this would be use-after-free

Reply:
Please note that these are two different variables: `prov_ctx` and `provctx` - confusing I agree but I did not want to make larger changes in the code.

Reply:
Oh wow, yes, that is absolutely horrible, I did not notice the `_`

Reply:
I think a rename of `provctx` (the argument to the function) to `p_prov_ctx` (or something similar) would only be a change on 2 lines, and as a first commit in this PR (to be squashed on merge) might make this function more maintainable in the future - WDYT? Agree out-of-scope for "fix leak", but as this is so bad it seems worth fixing on a "tidy up the kitchen a little bit each time" basis.

Presumably we wouldn't need to change the name of the parameter in the function declaration if we did that? (I mean: we don't from a C perspective, but if there's something in the CI that checks argument names match in declarations and definitions we might need to)

Reply:
I did it the other way around - renamed `prov_ctx`.